File: src/main/java/com/redbox/domain/article/service/ArticleService.java
package com.redbox.domain.article.service;

import com.redbox.domain.article.dto.ArticleResponse;
import com.redbox.domain.article.dto.CreateArticleRequest;
import com.redbox.domain.article.dto.UpdateArticleRequest;
import com.redbox.domain.article.entity.Article;
import com.redbox.domain.article.exception.ArticleNotFoundException;
import com.redbox.domain.article.repository.ArticleRepository;
import com.redbox.domain.user.service.UserService;
import com.redbox.global.entity.PageResponse;
import io.lettuce.core.RedisConnectionException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;

@Slf4j
@Service
@RequiredArgsConstructor
public class ArticleService {

    private final UserService userService;
    private final ArticleRepository articleRepository;
    private final RedisTemplate<String, Object> redisTemplate;
    private static final Duration CACHE_TTL = Duration.ofMinutes(30);       // 30분
    private static final String ARTICLE_PAGE_KEY = "article:page:%d";       //헌혈기사 1페이지 리스트

    @Transactional(readOnly = true)
    public PageResponse<ArticleResponse> getArticles(int page, int size) {
        Pageable pageable = PageRequest.of(page - 1, size, Sort.by("createdAt").descending());
        // 첫페이지만 캐싱
        if (page == 1) {
            String cacheKey = String.format(ARTICLE_PAGE_KEY, page);
            try {
                Object cached = redisTemplate.opsForValue().get(cacheKey);
                if (cached != null) {
                    return (PageResponse<ArticleResponse>) cached;
                }

                PageResponse<ArticleResponse> response = new PageResponse<>(articleRepository.findAll(pageable).map(ArticleResponse::new));
                redisTemplate.opsForValue().set(cacheKey, response, CACHE_TTL);
                return response;
            } catch (RedisConnectionException e) {
                log.error("Redis 연결 실패", e);
                return new PageResponse<>(articleRepository.findAll(pageable).map(ArticleResponse::new));
            }
        }

        // 첫페이지 외에는 DB로 가져옴
        return new PageResponse<>(articleRepository.findAll(pageable).map(ArticleResponse::new));
    }

    @Transactional
    public void createArticle(CreateArticleRequest request) {
        articleRepository.save(Article.builder()
                .userId(userService.getCurrentUserId())
                .subject(request.getSubject())
                .articleUrl(request.getUrl())
                .source(request.getSource())
                .build());

        redisTemplate.delete(String.format(ARTICLE_PAGE_KEY, 1));
    }

    @Transactional
    public void updateArticle(Long articleId, UpdateArticleRequest request) {
        Article article = articleRepository.findById(articleId)
                .orElseThrow(ArticleNotFoundException::new);

        article.updateArticle(request);
        redisTemplate.delete(String.format(ARTICLE_PAGE_KEY, 1));
    }

    public ArticleResponse getArticle(Long articleId) {
        Article article = articleRepository.findById(articleId)
                .orElseThrow(ArticleNotFoundException::new);

        return new ArticleResponse(article);
    }

    @Transactional
    public void deleteArticle(Long articleId) {
        Article article = articleRepository.findById(articleId)
                .orElseThrow(ArticleNotFoundException::new);

        articleRepository.delete(article);
        redisTemplate.delete(String.format(ARTICLE_PAGE_KEY, 1));
    }
}

File: migration/src/main/kotlin/com/redbox/domain/community/article/service/ArticleService.kt
package com.redbox.domain.community.article.service

import com.redbox.domain.community.article.dto.ArticleResponse
import com.redbox.domain.community.article.dto.CreateArticleRequest
import com.redbox.domain.community.article.dto.UpdateArticleRequest
import com.redbox.domain.community.article.entity.Article
import com.redbox.domain.community.article.exception.ArticleNotFoundException
import com.redbox.domain.community.article.repository.ArticleRepository
import com.redbox.global.auth.service.AuthenticationService
import com.redbox.global.entity.PageResponse
import io.lettuce.core.RedisConnectionException
import org.hibernate.query.sqm.tree.SqmNode.log
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Duration

@Service
class ArticleService(
    private val authenticationService: AuthenticationService,
    private val articleRepository: ArticleRepository,
    private val redisTemplate: RedisTemplate<String, Any>,
) {

    companion object {
        private val CACHE_TTL: Duration = Duration.ofMinutes(30)
        private const val ARTICLE_PAGE_KEY = "article:page:%d"
    }

    @Transactional(readOnly = true)
    fun getArticles(page: Int, size: Int): PageResponse<ArticleResponse> {
        val pageable = PageRequest.of(page - 1, size, Sort.by("createdAt").descending())

        // 첫페이지만 캐싱
        if (page == 1) {
            val cacheKey = String.format(ARTICLE_PAGE_KEY, page)
            return try {
                redisTemplate.opsForValue().get(cacheKey)?.let { cached ->
                    cached as PageResponse<ArticleResponse>
                } ?: run {
                    val response = PageResponse(
                        articleRepository.findAll(pageable).map { ArticleResponse.fromArticle(it) }
                    )
                    redisTemplate.opsForValue().set(cacheKey, response, CACHE_TTL)
                    response
                }
            } catch (e: RedisConnectionException) {
                log.error("Redis 연결 실패", e)
                PageResponse(articleRepository.findAll(pageable).map { ArticleResponse.fromArticle(it) })
            }
        }

        // 첫페이지 외에는 DB로 가져옴
        return PageResponse(articleRepository.findAll(pageable).map { ArticleResponse.fromArticle(it) })
    }

    @Transactional
    fun createArticle(request: CreateArticleRequest) {
        articleRepository.save(Article(null, authenticationService.getCurrentUserId(), request.subject!!,  request.url!!, request.source!!))
        //TODO: 기존 캐시 삭제
        redisTemplate.delete(String.format(ARTICLE_PAGE_KEY, 1));
    }

    @Transactional
    fun updateArticle(articleId: Long, request: UpdateArticleRequest) {
        val article = articleRepository.findById(articleId).orElseThrow {ArticleNotFoundException()}
        article.update(request)
        //TODO: 기존 캐시 삭제
        redisTemplate.delete(String.format(ARTICLE_PAGE_KEY, 1));
    }

    @Transactional
    fun deleteArticle(articleId: Long) {
        val article = articleRepository.findById(articleId).orElseThrow {ArticleNotFoundException()}

        articleRepository.delete(article)
        //TODO: 기존 캐시 삭제
        redisTemplate.delete(String.format(ARTICLE_PAGE_KEY, 1));
    }
}